src/libraries/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeName.cs

```diff
@@ -142,47 +142,111 @@ public string FullName
         {
             get
             {
-                // This is a recursive method over potentially hostile input. Protection against DoS is offered
-                // via the [Try]Parse method and TypeNameParserOptions.MaxNodes property at construction time.
-                // This FullName property getter and related methods assume that this TypeName instance has an
-                // acceptable node count.
-                //
-                // The node count controls the total amount of work performed by this method, including:
-                // - The max possible stack depth due to the recursive methods calls; and
-                // - The total number of bytes allocated by this function. For a deeply-nested TypeName
-                //   object, the total allocation across the full object graph will be
-                //   O(FullName.Length * GetNodeCount()).
-
-                if (_fullName is null)
+                if (_fullName is not null) // Fast path for already computed FullName and nested types
                 {
-                    if (IsConstructedGenericType)
+                    if (_nestedNameLength > 0 && _fullName.Length > _nestedNameLength) // Declaring types
                     {
-                        _fullName = TypeNameParserHelpers.GetGenericTypeFullName(GetGenericTypeDefinition().FullName.AsSpan(),
-#if SYSTEM_PRIVATE_CORELIB
-                            CollectionsMarshal.AsSpan(_genericArguments));
-#else
-                            _genericArguments.AsSpan());
-#endif
+                        // Stored fullName represents the full name of the nested type.
+                        // Example: Namespace.Declaring+Nested
+                        _fullName = _fullName.Substring(0, _nestedNameLength);
```

Comment:
How about thread safety of this? Concurrent initialization with the same content should be safe, but this branch is modifying to different content.

Reply:
`TypeName` is not thread safe. In worst case, multiple different threads are going to perform a string allocation.

Reply:
> `TypeName` is not thread safe.

Isn't it? It's immutable. There might be some extra allocations but it would work fine when accessed from multiple threads, wouldn't it?

Reply:
You are right, it's fine to assume that something that is immutable is also thread-safe.

And again, in worst case, multiple threads would allocate the same string and duplicate the work, but each of them would end up computing the same text.